src/main/java/org/jenkinsci/plugins/mesos/MesosCloud.java

```diff
@@ -65,13 +88,36 @@ public MesosCloud(String name) throws InterruptedException, ExecutionException {
    *
    * @return A future reference to the launched node.
    */
-  private CompletableFuture<Node> startAgent() {
-    // mesos.enqueueAgent().thenCompose(mesosSlave -> mesosSlave.waitUntilOnlineAsync());
-    throw new NotImplementedException();
-  }
-
   @Override
   public boolean canProvision(Label label) {
-    throw new NotImplementedException();
+    // TODO: implement executor limits
+    return true;
+  }
+
+  public MesosApi getMesosClient() {
+    return this.mesos;
+  }
+
+  /**
+   * Start a Jenkins agent.jar on Mesos.
+   *
+   * <p>Provide a callback for Jenkins to start a Node.
+   *
+   * @return A future reference to the launched node.
+   */
+  public Future<Node> startAgent() throws Exception {
+    return mesos
+        .enqueueAgent(this, 0.1, 32)
+        .thenCompose(
+            mesosSlave -> {
+              try {
+                Jenkins.getInstanceOrNull().addNode(mesosSlave);
```

Comment:
@jeschkies believe adding nodes to the jenkins instance will persist the serialized structure to disk, since the slaves have podIds tied to slaves this can act as our persistence layer thoughts?

Reply:
I was wondering the same. The Nomad plugin uses ephemeral nodes. We have to research that.